### PR TITLE
allow users to define custom effects

### PR DESCRIPTION
**Pull request by fwbrasil**
The new `Tag` solved some safety issues when defining new effects. I think we'll still need to improve the usability of the APIs but I think it's in a state where we can allow users to start using it.

### CODE REVIEW

File: kyo-core/shared/src/main/scala/kyo/core.scala [hunk 1]
```diff
@@ -104,6 +104,26 @@ object core:
                         r.asInstanceOf[Result[T]]
             handleLoop(state, value)
         end handle
+
+        inline def deepHandle[Command[_], S, T: Flat](
```
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/core.scala:108]: Moved to a public scope.

File: README.md [hunk 2]
```diff
@@ -2100,6 +2100,104 @@ For further examples, Kyo's [example ledger service](https://github.com/getkyo/k
...
+    Counter.dec(2)(10)
+  }
+
+// Typically wrapped in a `run`
+// method by Kyo effects
+val b: Int < Any = 
+  Counter.handle(handler)((), a)
+```
+
+In this example, `a` is a computation prints `5` and `2` and finally returns the value `10`. We handle the `Counter` effect using `Counter.run`, which returns a tuple containing the final count and the result of the computation.
```
Comment (hearnadam) [inline README.md:2156]: Maybe we can just show the console output.

File: README.md [hunk 3]
```diff
@@ -2100,6 +2100,104 @@ For further examples, Kyo's [example ledger service](https://github.com/getkyo/k
...
+```scala mdoc:nest
+val a: Int < Counter = 
+  Counter.inc(5) {
+    Counter.dec(2)(10)
+  }
+
+// Typically wrapped in a `run`
+// method by Kyo effects
+val b: Int < Any = 
+  Counter.handle(handler)((), a)
```
Comment (hearnadam) [inline README.md:2153]: Perhaps we can define a run method? that way users can understand better the goal.

File: README.md [hunk 4]
```diff
@@ -2100,6 +2100,104 @@ For further examples, Kyo's [example ledger service](https://github.com/getkyo/k
...
+```
+
+The `inc` and `dec` operations suspend the `Counter` effect with the provided amount and the computation `v`. The `inc` operation increments the counter by the specified amount, while the `dec` operation decrements it.
+
+### Effect Handlers
+
+Effect handlers specify how to interpret and execute the operations of an effect. In Kyo, it's a common style pattern to define effect handling using a `run` method in the effect's companion object. Here's an example of defining the run method and the effect handler for the `Counter` effect that prints commands to the console as an example:
+
+```scala mdoc:nest
+val handler = new Handler[[T] =>> (Int, T), Counter, Any]:
```
Comment (hearnadam) [inline README.md:2135]: Which object should this go in?

Should we encourage users to model handlers via extension?

**Comment by fwbrasil** [comment]
I'm closing this for now since we're looking into core improvements https://github.com/getkyo/kyo/issues/462

**Pull request closed without merging**